src/registrar/views/domain.py

```diff
@@ -221,6 +241,300 @@ def form_valid(self, formset):
+            initial_data.append({})
+
+        return initial_data
+
+    def get_success_url(self):
+        """Redirect to the DS Data page for the domain."""
+        return reverse("domain-dns-dnssec-dsdata", kwargs={"pk": self.object.pk})
+
+    def get_context_data(self, **kwargs):
+        """Adjust context from FormMixin for formsets."""
+        context = super().get_context_data(**kwargs)
+        # use "formset" instead of "form" for the key
+        context["formset"] = context.pop("form")
+
+        # set the dnssec_ds_confirmed flag in the context for this view
+        # based either on the existence of DS Data in the domain,
+        # or on the flag stored in the session
+        domain = self.get_object()
+        dnssecdata: extensions.DNSSECExtension = domain.dnssecdata
+
+        if dnssecdata is not None and dnssecdata.dsData is not None:
+            self.request.session["dnssec_ds_confirmed"] = True
+
+        context["dnssec_ds_confirmed"] = self.request.session.get(
+            "dnssec_ds_confirmed", False
+        )
+        return context
+
+    def post(self, request, *args, **kwargs):
+        """Formset submission posts to this view."""
+        self.object = self.get_object()
+        formset = self.get_form()
+
+        if "confirm-ds" in request.POST:
+            request.session["dnssec_ds_confirmed"] = True
+            request.session["dnssec_key_confirmed"] = False
+            return super().form_valid(formset)
+
+        if "btn-cancel-click" in request.POST:
+            return redirect("/", {"formset": formset}, RequestContext(request))
+
+        if formset.is_valid():
+            return self.form_valid(formset)
+        else:
+            return self.form_invalid(formset)
+
+    def form_valid(self, formset):
+        """The formset is valid, perform something with it."""
+
+        # Set the dnssecdata from the formset
+        dnssecdata = extensions.DNSSECExtension()
+
+        for form in formset:
+            try:
+                # if 'delete' not in form.cleaned_data
+                # or form.cleaned_data['delete'] == False:
+                dsrecord = {
+                    "keyTag": form.cleaned_data["key_tag"],
+                    "alg": int(form.cleaned_data["algorithm"]),
+                    "digestType": int(form.cleaned_data["digest_type"]),
+                    "digest": form.cleaned_data["digest"],
+                }
+                if dnssecdata.dsData is None:
+                    dnssecdata.dsData = []
+                dnssecdata.dsData.append(common.DSData(**dsrecord))
+            except KeyError:
+                # no server information in this field, skip it
+                pass
+        domain = self.get_object()
+        try:
+            domain.dnssecdata = dnssecdata
+        except RegistryError as err:
+            errmsg = "Error updating DNSSEC data in the registry."
+            logger.error(errmsg)
+            logger.error(err)
```

Comment:
**comment** actualy like that you log both here, nice idea!

Reply:
thanks